api/resolvers/vault.js

```diff
@@ -29,12 +41,21 @@ export default {
         }))
       }
 
-      for (const entry of entries) {
-        txs.push(models.vaultEntry.update({
-          where: { userId_key: { userId: me.id, key: entry.key } },
-          data: { value: entry.value, iv: entry.iv }
-        }))
+      const wallets = await models.wallet.findMany({ where: { userId: me.id } })
```

Comment:
I think fetching the wallets outside of the transaction is fine since it's a read-only operation.

Else this needs to become an interactive transaction :eyes:

Reply:
Regardless of whether this is interactive, this code is/was vulnerable to race conditions if a wallet is upserted while the vault key is updated.

Reply:
Okay, I assume this is blocking this PR

Reply:
Fixed in 72ac2225

Reply:
This wasn't blocking because even after your changes the code is vulnerable to a race "if a wallet is upserted while the vault key is updated." Your change *only* protects against hash change races.

e.g. 

tx1: read wallets (wallets with ids [1,2])
tx2: insert wallet (wallet [3])
tx2: commit
tx1: update hash
tx1: commit

Wallet [3] is broken because it's encrypted with a different key.

Reply:
The above race is very unlikely to occur, but that's why it wasn't blocking. The fix is nontrivial too: the only way to afaict is to give each vault entry a foreign key constraint on the correct key hash. That way, if it's updated any concurrent inserts would fail.

Reply:
Both transactions read the old vault key hash first, so one of them should fail when they try to update it at the end because they won't find the row anymore, no?